### PR TITLE
Feature/game config

### PR DESCRIPTION
**Pull request by D3strukt0r**

### CODE REVIEW

File: client/src/main/java/jass/client/eventlistener/BroadcastGameModeEventListener.java [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+import jass.lib.message.BroadcastGameModeData;
+
+/**
+ * @author Manuele Vaccari
+ * @version %I%, %G%
+ * @since 0.0.1
+ */
+public interface BroadcastGameModeEventListener {
+    /**
+     * Executes when a user is requested to choose a game mode.
```
Comment (jokerengine) [inline client/src/main/java/jass/client/eventlistener/BroadcastGameModeEventListener.java:12]: Executes when a user has chosen a game mode.

File: client/src/main/java/jass/client/controller/GameController.java [hunk 2]
```diff
@@ -675,4 +625,24 @@ public void updateUserNames() {
...
+    @Override
+    public void onBroadcastGameMode(final BroadcastGameModeData data) {
+        Platform.runLater(() -> {
+            // TODO Make this more beautiful
+            if (data.getGameMode() == GameMode.TRUMPF) {
+                mode.setText("Mode: " + data.getGameMode().toString() + " | Card: " + data.getTrumpfSuit());
+            } else {
+                mode.setText("Mode: " + data.getGameMode().toString());
+            }
+            // TODO Maybe enable buttons in here to start game?
```
Comment (jokerengine) [inline client/src/main/java/jass/client/controller/GameController.java]: I think the buttons should be enabled here, but only when it's the person's turn. I think for now just activate them and I'll add some logic checking if it's that players turn. Also don't forget to set the game mode in the GameUtil.

File: client/src/main/java/jass/client/util/GameUtil.java [hunk 3]
```diff
@@ -1,27 +1,38 @@
...
 import org.apache.logging.log4j.LogManager;
 import org.apache.logging.log4j.Logger;
 import jass.lib.servicelocator.Service;
 
 import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
 
 /**
  * @author Victor Hargrave
```
Comment (jokerengine) [inline client/src/main/java/jass/client/util/GameUtil.java]: you can add yourself as an author too :P

File: client/src/main/java/jass/client/util/SocketUtil.java [hunk 4]
```diff
@@ -256,54 +268,88 @@ public boolean isLoggedIn() {
...
+        broadcastGameModeListener.add(listener);
     }
 
     /**
      * @param msgType Message to send to listener listening to "game-found"
      *                event.
      * @param msgData The message to send to the listeners.
      *
-     * @author Thomas Weber
+     * @author Thomas Weber, Manuele Vaccari
```
Comment (jokerengine) [inline client/src/main/java/jass/client/util/SocketUtil.java]: add me too? :*)

File: server/src/main/java/jass/server/Main.java [hunk 5]
```diff
@@ -127,8 +127,9 @@ public static void main(final String[] args) {
         logger.info("SSL is " + (secure ? "enabled" : "disabled"));
 
         // Initialize Other Util Classes and add to ServiceLocator
-        SearchGameUtil sGU = new SearchGameUtil();
-        ServiceLocator.add(sGU);
+        SearchGameUtil searchGameUtil = new SearchGameUtil();
+        searchGameUtil.start();
```
Comment (jokerengine) [inline server/src/main/java/jass/server/Main.java:131]: I haven't seen the rest yet, but already like how this looks. :D

File: server/src/main/java/jass/server/util/GameUtil.java [hunk 6]
```diff
@@ -0,0 +1,160 @@
...
+        UserEntity playerOne = clientPlayerOne.getUser();
+        this.clientPlayerTwo = clientPlayerTwo;
+        UserEntity playerTwo = clientPlayerTwo.getUser();
+        this.clientPlayerThree = clientPlayerThree;
+        UserEntity playerThree = clientPlayerThree.getUser();
+        this.clientPlayerFour = clientPlayerFour;
+        UserEntity playerFour = clientPlayerFour.getUser();
+
+        // Add event listeners
+        clientPlayerOne.addChosenGameModeEventListener(this);
```
Comment (jokerengine) [inline server/src/main/java/jass/server/util/GameUtil.java:85]: in the first round, player one gets to pick yes. But.. we don't decide based on the fact that this is player one, but instead should look at the round table and which user id gets to pick for this round.

File: server/src/main/java/jass/server/util/GameUtil.java [hunk 7]
```diff
@@ -0,0 +1,160 @@
...
+
+        // Send a message to player one to choose a game mode
+        ChooseGameMode chooseGameMode = new ChooseGameMode(new ChooseGameModeData());
+        clientPlayerOne.send(chooseGameMode);
+    }
+
+    @Override
+    public void onChosenGameMode(final ChosenGameModeData data) {
+        // Check with token if player one actually is the one who sent the data.
+        if (data.getToken().equals(clientPlayerOne.getToken())) {
```
Comment (jokerengine) [inline server/src/main/java/jass/server/util/GameUtil.java]: in round 2 clientPlayerTwo gets to pick the game mode. This case needs to be handled too.

File: server/src/main/java/jass/server/util/GameUtil.java [hunk 8]
```diff
@@ -0,0 +1,160 @@
...
+                broadcastGameMode = new BroadcastGameMode(new BroadcastGameModeData(gameMode));
+            }
+            broadcast(broadcastGameMode);
+        }
+    }
+
+    /**
+     * @param message Message to broadcast to all players.
+     */
+    private void broadcast(final Message message) {
```
Comment (jokerengine) [inline server/src/main/java/jass/server/util/GameUtil.java:163]: I like..much better handled :P

File: server/src/main/java/jass/server/util/SearchGameUtil.java [hunk 9]
```diff
@@ -19,35 +19,32 @@
...
 /**
  * Creates a game once 4 clients are searching for a game in the lobby.
  *
- * @author Thomas Weber
+ * @author Thomas Weber, Manuele Vaccari
  * @version %I%, %G%
  * @since 0.0.1
  */
-public final class SearchGameUtil implements Service {
+public final class SearchGameUtil extends Thread implements Service, Closeable {
```
Comment (jokerengine) [inline server/src/main/java/jass/server/util/SearchGameUtil.java:41]: really nice job. Much cleaner

File: server/src/main/java/jass/server/util/GameUtil.java [hunk 10]
```diff
@@ -0,0 +1,160 @@
...
+        clientPlayerOne.send(chooseGameMode);
+    }
+
+    @Override
+    public void onChosenGameMode(final ChosenGameModeData data) {
+        // Check with token if player one actually is the one who sent the data.
+        if (data.getToken().equals(clientPlayerOne.getToken())) {
+            GameMode gameMode = data.getGameMode();
+            BroadcastGameMode broadcastGameMode;
+            if (gameMode == GameMode.TRUMPF) {
```
Comment (jokerengine) [inline server/src/main/java/jass/server/util/GameUtil.java]: we should save the game mode to the database.

**Review by jokerengine — COMMENTED** [review]
in general, really great work. Refactoring looks good. 
don't forget we're saving everything to our db and that the player who chooses the round is not necessarily player one, but the player id selected in the round table.

**Review by sasatrajkova — APPROVED** [review]

**Pull request merged**